lib/mayaUsd/ufe/Utils.cpp

```diff
@@ -80,19 +82,30 @@ bool stringBeginsWithDigit(const std::string& inputString)
     return false;
 }
 
-// find position index for a given layer in the local LayerStack ( strong-to-weak order )
-uint32_t findLayerIndex(const UsdStageWeakPtr& stage, const PXR_NS::SdfLayerHandle& layer)
+// This function calculates the position index for a given layer across all
+// the site's local LayerStack.
+uint32_t findLayerIndex(const UsdPrim& prim, const PXR_NS::SdfLayerHandle& layer)
 {
-    const auto& layerStack = stage->GetLayerStack();
-
     uint32_t position { 0 };
 
-    auto iter = std::find(std::begin(layerStack), std::end(layerStack), layer);
+    const PXR_NS::PcpPrimIndex& primIndex = prim.GetPrimIndex();
+
+    // iterate through the expanded primIndex
+    for (const PcpNodeRef& node : primIndex.GetNodeRange()) {
+
+        TF_AXIOM(node);
 
-    assert(iter != layerStack.end());
+        const PcpLayerStackSite&   site = node.GetSite();
+        const PcpLayerStackRefPtr& layerStack = site.layerStack;
 
-    if (iter != layerStack.end()) {
-        return std::distance(layerStack.begin(), iter);
+        // iterate through the "local" Layer stack for each site
+        // to find the layer
+        for (SdfLayerRefPtr const& l : layerStack->GetLayers()) {
+            if (l == layer) {
+                return position;
+            }
+            ++position;
+        }
```

Comment:
The original algorithm that I wrote here was very naive and didn't handle the cases where the attribute came in across composition arcs. By walking the expanded primIndex, we would be able to visit all the "**sites**" and their "**local layer stack**".

I am also thinking of another way to write this code but believe for the pass should be good. 

Thanks to Brett Levin and Spiff to help me better understand the "primindex" and "site" concepts.

Reply:
So this linearizes the complete layer stack and expands out each site, such that if there were m node sites each with n layers, then the value of position at the end of the loop would be (m x n) - 1, right?